src/mem.rs

```diff
@@ -1,30 +1,41 @@
-use crate::src::error::Rav1dError::ENOMEM;
-use crate::src::error::Rav1dResult;
 use libc::free;
-use libc::malloc;
 use libc::posix_memalign;
-use libc::pthread_mutex_destroy;
-use libc::pthread_mutex_init;
-use libc::pthread_mutex_lock;
-use libc::pthread_mutex_t;
-use libc::pthread_mutex_unlock;
-use libc::pthread_mutexattr_t;
-use libc::uintptr_t;
-use std::ffi::c_int;
+use std::collections::LinkedList;
 use std::ffi::c_void;
+use std::sync::Mutex;
 
-#[repr(C)]
-pub struct Rav1dMemPool {
-    pub lock: pthread_mutex_t,
-    pub buf: *mut Rav1dMemPoolBuffer,
-    pub ref_cnt: c_int,
-    pub end: c_int,
+pub struct MemPool<T> {
+    bufs: Mutex<LinkedList<Vec<T>>>,
 }
 
-#[repr(C)]
-pub struct Rav1dMemPoolBuffer {
-    pub data: *mut c_void,
-    pub next: *mut Rav1dMemPoolBuffer,
+impl<T> MemPool<T> {
```

Comment:
For `picture_pool`, it could be hardcoded to `u8`, but if we later add pooling back to the other fields, some of them use non-`u8`s.  And since it's trivial to make this generic, I didn't see a reason why not to.